src/site/antora/modules/ROOT/pages/manual/plugins.adoc

```diff
@@ -14,88 +14,274 @@
-
-[#attribute-types]
-=== Attribute types
-
-* PluginAttribute: The parameter must be convertible from a String using a TypeConverter. Most built-in types are already supported, but custom TypeConverter plugins may also be provided for more type support. Note that PluginBuilderAttribute can be used in builder class fields as an easier way to provide default values.
-* PluginElement: The parameter may represent a complex object that itself has parameters that can be configured. This also supports injecting an array of elements.
-* PluginConfiguration: The current Configuration object will be passed to the plugin as a parameter.
-* PluginNode: The current Node being parsed will be passed to the plugin as a parameter.
-* PluginValue: The value of the current Node or its attribute named value.
-
-[#constraint-validators]
-=== Constraint validators
+Gradle::
++
+[source,groovy,subs="+attributes"]
+----
+dependencies {
+  // Process sources using `log4j-core` providing `org.apache.logging.log4j.core.config.plugins.processor.PluginProcessor` that generates `Log4j2Plugins.dat` -->
+  annotationProcessor('org.apache.logging.log4j:log4j-core:{log4j-core-version}')
+}
+----
+====
 
-Plugin factory fields and parameters can be automatically validated at runtime using constraint validators inspired by the Bean Validation spec. The following annotations are bundled in Log4j, but custom ConstraintValidators can be created as well.
+[#plugin-discovery]
+== Discovering plugins
 
-* Required: This annotation validates that a value is non-empty. This covers a check for null as well as several other scenarios: empty CharSequence objects, empty arrays, empty Collection instances, and empty Map instances.
-* ValidHost: This annotation validates that a value corresponds to a valid hostname. This uses the same validation as InetAddress::getByName.
-* ValidPort: This annotation validates that a value corresponds to a valid port number between 0 and 65535.
+link:../javadoc/log4j-core/org/apache/logging/log4j/core/config/plugins/util/PluginManager.html[`PluginManager`] is responsible for discovering plugins and loading their descriptions.
+It locates plugins by looking in following places in given order:
 
-[#key-providers]
-== Key providers
+. Plugin descriptor files on the classpath.
+These files are generated automatically at compile-time by the Log4j plugin annotation processor.
+See <<plugin-registry>> for details.
 
-Some components within Log4j may provide the ability to perform data encryption. These components require a secret key to perform the encryption. Applications may provide the key by creating a class that implements the SecretKeyProvider interface.
+. *[OSGi only]* Serialized plugin listing files in each active OSGi bundle.
+A `BundleListener` is added on activation to continue checking new bundles after Log4j Core has started.
```

Comment:
Should we say that it is the same `Log4j2Plugins.dat` file that we use?

Reply:
I prefer not to – see [this comment](https://github.com/apache/logging-log4j2/pull/2696#discussion_r1663730329).